### ISSUE TITLE
PyInstaller uses amd64 DLLs to satisfy x86 dependencies

### ISSUE THREAD

**Issue by pyinstaller-tickets-migration** [pyinstaller/pyinstaller#802]
**Original date**: 2013/10/17
**Original reporter**: *torsten DOT landschoff AND dynamore DOT de*

We build a amd64 binary with PyInstaller which worked fine. Somehow it does not work anymore. I was able to trace this to the fact that we installed GraphViz on the build machine for generating some graphs inside documentation, which includes x86 DLLs for libglib etc. which are used by PyInstaller instead of the correct amd64 DLLs installed elsewhere.

The following steps will reproduce this:

## Prepare Python
Install PyGTK like described here: http://stackoverflow.com/questions/14567921/python-pygtk-windows-7-64-bit#14587144
 - Install Python 2.7.5 x64 from python.org
 - Install GTK+ Bundle from http://ftp.gnome.org/pub/GNOME/binaries/win64/gtk+/2.22/gtk+-bundle_2.22.1-20101229_win64.zip - see http://www.gtk.org/download/win64.php for some infos. Install/unzip to c:\opt\gtk
 - Install PyGObject from http://www.lfd.uci.edu/~gohlke/pythonlibs/bnrm5n67/pygobject-2.28.6.win-amd64-py2.7.exe (linked from http://www.lfd.uci.edu/~gohlke/pythonlibs/#pygtk)

## Install GraphViz
Download and install GraphViz from http://www.graphviz.org/pub/graphviz/stable/windows/graphviz-2.30.1.msi

## Create a trivial example using glib
Create a trivial Python script that just shows that glib can be loaded:
```
#!python
import glib

if __name__ == "__main__":
    version_string = ".".join(map(str, glib.glib_version))
    print "Successfully loaded glib version %s" % version_string
```

This runs fine from the Windows command line like this:
```
> cd %HOMEPATH%\dllarch
> set PATH=%PATH%;c:\Python27;c:\opt\gtk\bin
> python dllarch.py
Successfully loaded glib version 2.26.1
```

## Create .EXE with PyInstaller
Now use PyInstaller to create an executable:
```
> python ..\pyinstaller\pyinstaller.py --log-level DEBUG -y dllarch.py > pyinstaller.log 2>&1
```
The generated executable seems to work:
```
> dist\dllarch\dllarch.exe
Successfully loaded glib version 2.26.1
```
But checking the dependencies of the _glib extension module with [DependencyWalker](http://www.dependencywalker.com/) shows that the result will not actually work outside the development environment:
```
> ..\Desktop\depends.exe dist\dllarch\glib._glib.pyd
```
[[Image(screen_114.png)]]
[[Image(screen_115.png)]]

Looking at attachment:out00-Analysis.toc one can see that PyInstaller actually included the DLLs from the GraphViz installation instead of the correct one from c:\opt\gtk. Renaming the GTK installation reveals that the executable created by PyInstaller is broken:

```
> rename c:\opt\gtk gtk-hidden
> dist\dllarch\dllarch.exe
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\Torsten\pyinstaller\PyInstaller\loader\pyi_importers.py", line
 270, in load_module
    exec(bytecode, module.__dict__)
  File "C:\Users\Torsten\dllarch\build\dllarch\out00-PYZ.pyz\glib", line 22, in
 <module>
  File "C:\Users\Torsten\pyinstaller\PyInstaller\loader\pyi_importers.py", line
 414, in load_module
    module = imp.load_module(fullname, fp, filename, self._c_ext_tuple)
ImportError: DLL load failed: %1 is not a valid Win32 application.
```

Tested with revision r096d4967471b3f8643222dabff5e01a5ce127c18 of PyInstaller.

**Comment by codewarrior0** [comment]
I believe the solution is to have `PyInstaller.depend.bindepend.getfullnameof()` on Windows inspect the PE file headers of any DLLs found and ignore ones whose machine type (`pe.FILE_HEADER.Machine`) doesn't match the one we're building for.

**Comment by matysek** [comment]
> on Windows inspect the PE file headers of any DLLs

This sounds feasible.

**Comment by codewarrior0** [comment]
Should be fixed in 574f04b65ed977e4cf77efc2ed912bff06a33d4e
